BJS2-77984 become mentor

Fagan89 opened the pull request:
Краткое описание: Необходимо реализовать основные операции: получить всех менти одного пользователя, получить всех менторов одного пользователя, удалить менти и удалить ментора.

Commit:
BJS2-77984
1.добавил Dto
2.добавил Mapper-ры
3.добавил Service и Collector

Commit:
BJS2-77984
1.добавил Валидацию

Commit:
Merge branch 'hippogriff-master-stream10' into feature/BJS2-77984-become-mentor

Commit:
BJS2-77984

Commit:
BJS2-77984
Исправил каталог

FROLOV-KI commented on src/main/java/school/faang/user_service/controller/mentorship/MentorshipController.java:
```diff
@@ -0,0 +1,47 @@
+public class MentorshipController {
+    private final MentorshipService mentorshipService;
+
+    @GetMapping("/mentorship/mentees/{userId}/search")
```
/search лишнее

FROLOV-KI commented on src/main/java/school/faang/user_service/controller/mentorship/MentorshipController.java:
```diff
@@ -0,0 +1,47 @@
+        return ResponseEntity.ok(mentees);
+    }
+
+    @GetMapping("/mentorship/mentors{userId}/search")
```
mentors/{userId} пропустил / между + /search лишнее

FROLOV-KI commented on src/main/java/school/faang/user_service/controller/mentorship/MentorshipController.java:
```diff
@@ -0,0 +1,47 @@
+        return ResponseEntity.ok(mentors);
+    }
+
+    @DeleteMapping("/mentorship/{mentorId}/mentee/delete/{menteeId}")
```
/delete лишнее, сам метод говорит, что за действие
я бы наверное сделал то-то типа /mentorship/mentors/{mentorId}/{menteeId}

FROLOV-KI commented on src/main/java/school/faang/user_service/controller/mentorship/MentorshipController.java:
```diff
@@ -0,0 +1,47 @@
+        return ResponseEntity.ok().build();
+    }
+
+    @DeleteMapping("/mentorship/{menteeId}/mentor/delete/{mentorId}")
```
/delete лишнее, сам метод говорит, что за действие
я бы наверное сделал то-то типа /mentorship/mentees/{menteeId}/{mentorId}

FROLOV-KI commented on src/main/java/school/faang/user_service/controller/mentorship/MentorshipController.java:
```diff
@@ -0,0 +1,47 @@
+    @DeleteMapping("/mentorship/{menteeId}/mentor/delete/{mentorId}")
+    public ResponseEntity<Void> deleteMentor(@PathVariable long mentorId,
+                                             @PathVariable long menteeId) {
+        mentorshipService.deleteMentor(mentorId, menteeId);
```
А менторов может быть много? Может быть тут будет достаточно только id подопечного?

Fagan89 replied:
да, у пользователя может быть несколько менторов

FROLOV-KI commented on src/main/java/school/faang/user_service/service/mentorship/MentorshipService.java:
```diff
@@ -0,0 +1,87 @@
+
+    private User findUserIntoMentorshipRepository(long userId) {
+        return mentorshipRepository.findById(userId)
+                .orElseThrow(() -> new IllegalArgumentException("Такого пользователя нет в базе"));
```
IllegalArgumentException сюда словно бы не подходит, параметр передан не неверный, просто такой записи в бд нет. Сюда бы лучше кастомный exception, что-то типа NoSuchMentorException или NoSuchMentorshipException. А то написано "Такого пользователя нет в базе", а в какой не понятно.

FROLOV-KI commented on src/main/java/school/faang/user_service/service/mentorship/MentorshipService.java:
```diff
@@ -0,0 +1,87 @@
+
+    private User findUserIntoUserRepository(long userId) {
+        return userRepository.findById(userId)
+                .orElseThrow(() -> new IllegalArgumentException("Такого пользователя нет в базе"));
```
+

FROLOV-KI commented on src/main/java/school/faang/user_service/validation/mentorship/MentorshipValidation.java:
```diff
@@ -0,0 +1,52 @@
+
+    public static void validateMenteesNonEmpty(List<User> mentees) {
+        if (mentees == null || mentees.isEmpty()) {
+            throw new IllegalArgumentException("У пользователя нету учеников");
```
+

FROLOV-KI commented on src/main/java/school/faang/user_service/validation/mentorship/MentorshipValidation.java:
```diff
@@ -0,0 +1,52 @@
+
+    public static void validateMentorsNonEmpty(List<User> mentors) {
+        if (mentors == null || mentors.isEmpty()) {
+            throw new IllegalArgumentException("У пользователя нету наставников");
```
+

FROLOV-KI commented on src/main/java/school/faang/user_service/validation/mentorship/MentorshipValidation.java:
```diff
@@ -0,0 +1,52 @@
+
+    public static void validateMentorshipUsers(User mentor, User mentee) {
+        if (mentor == null || mentee == null) {
+            throw new IllegalArgumentException("Ментор и ученик не могут быть null");
```
+

Commit:
BJS2-77984
Скорректировал исключения

20cognosce commented on src/main/java/school/faang/user_service/mapper/mentorship/MentorsMapper.java:
```diff
@@ -0,0 +1,17 @@
+import school.faang.user_service.dto.mentorship.MentorDto;
+import school.faang.user_service.entity.User;
+
+public class MentorsMapper {
```
Можно было один маппер вместо двух

20cognosce commented on src/main/java/school/faang/user_service/mapper/mentorship/MenteesMapper.java:
```diff
@@ -0,0 +1,17 @@
+
+public class MenteesMapper {
+
+    @Builder
```
над самим классом DTO / Entity можно вешать

20cognosce commented on src/main/java/school/faang/user_service/service/mentorship/MentorshipService.java:
```diff
@@ -0,0 +1,88 @@
+    private final MentorshipRepository mentorshipRepository;
+    private final UserRepository userRepository;
+
+    @Transactional
```
readOnly

20cognosce commented on src/main/java/school/faang/user_service/validation/mentorship/MentorshipValidation.java:
```diff
@@ -0,0 +1,53 @@
+
+public class MentorshipValidation {
+
+    public static void validateMenteesNonEmpty(List<User> mentees) {
+        if (mentees == null || mentees.isEmpty()) {
+            throw new DataValidationException("У пользователя нету учеников");
+        }
+    }
+
+    public static void validateMentorsNonEmpty(List<User> mentors) {
+        if (mentors == null || mentors.isEmpty()) {
+            throw new DataValidationException("У пользователя нету наставников");
+        }
+    }
```
ну в случае если нет менторов или менти не надо ошибку кидать, это нормальное состояние
пустой список возвращать нормально

20cognosce commented on src/main/java/school/faang/user_service/service/mentorship/MentorshipService.java:
```diff
@@ -0,0 +1,88 @@
+                .toList();
+    }
+
+    @Transactional
```
readOnly

20cognosce commented on src/main/java/school/faang/user_service/service/mentorship/MentorshipService.java:
```diff
@@ -0,0 +1,88 @@
+
+        MentorshipValidation.validateMentorsNonEmpty(mentors);
+
+        return mentors.stream()
+                .map(MentorsMapper::toDto)
+                .toList();
```
сервис должен возвращать сущности, не надо тут делать зависимость на маппер
в фасад вынести маппинг

20cognosce commented on src/main/java/school/faang/user_service/service/mentorship/MentorshipService.java:
```diff
@@ -0,0 +1,88 @@
+        validateMentorshipUsers(mentor, mentee);
+        validateIsMentorOf(mentor, mentee);
+
+        mentee.getMentors().remove(mentor);
+        userRepository.save(mentor);
```
что ты удалил и что ты сохранил
это не будет работать

Commit:
BJS2-77984
1. добавил Фасад к сервису
2. убрал валидацию пустого списка
3. Маппер объединил
4. сообщения исключений изменил на английский

Commit:
Merge branch 'hippogriff-master-stream10' into feature/BJS2-77984-become-mentor

Commit:
BJS2-77984
1. добавил Unit тесты для фасада сервиса

Fagan89 commented:
Исправил замечания, начал покрывать тестами в этой ветке, продолжу в другой

20cognosce commented on src/test/java/school/faang/user_service/service/mentorship/MentorshipServiceFacadeTest.java:
```diff
@@ -0,0 +1,53 @@
+    @Mock
+    private MentorshipService mentorshipService;
+
+    @InjectMocks
+    private MentorshipServiceFacade mentorshipServiceFacade;
```
на фасад зачем тесты писать если там просто маппер работает и всё

20cognosce commented on src/main/java/school/faang/user_service/controller/mentorship/MentorshipController.java:
```diff
@@ -0,0 +1,49 @@
+@RequiredArgsConstructor
+@RequestMapping("/api/v1/users")
+public class MentorshipController {
+    private final MentorshipService mentorshipService;
+    private final MentorshipServiceFacade mentorshipServiceF;
```
не надо нести две зависимости, зачем
фасад должен полностью оборачивать сервис
